File: app/src/main/java/com/pr0gramm/app/ui/base/BaseDialogFragment.java
package com.pr0gramm.app.ui.base;

import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.os.Bundle;
import android.support.v4.app.FragmentActivity;

import com.f2prateek.dart.Dart;
import com.pr0gramm.app.ActivityComponent;
import com.pr0gramm.app.Dagger;
import com.pr0gramm.app.ui.dialogs.DialogDismissListener;
import com.trello.rxlifecycle.LifecycleTransformer;
import com.trello.rxlifecycle.components.support.RxAppCompatDialogFragment;

import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * A robo fragment that provides lifecycle events as an observable.
 */
public abstract class BaseDialogFragment extends RxAppCompatDialogFragment {
    private Unbinder unbinder;

    public final <T> LifecycleTransformer<T> bindToLifecycleAsync() {
        return new AsyncLifecycleTransformer<>(bindToLifecycle());
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        injectComponent(Dagger.activityComponent(getActivity()));

        if (getArguments() != null)
            Dart.inject(this, getArguments());

        super.onCreate(savedInstanceState);
    }

    protected abstract void injectComponent(ActivityComponent activityComponent);

    @Override
    public void onStart() {
        super.onStart();

        // bind dialog. It is only created in on start.
        Dialog dialog = getDialog();
        if (dialog != null) {
            unbinder = ButterKnife.bind(this, dialog);
            onDialogViewCreated();
        }
    }

    protected void onDialogViewCreated() {
    }

    @Override
    public void onDestroyView() {
        if (getDialog() != null && getRetainInstance())
            getDialog().setDismissMessage(null);

        super.onDestroyView();

        if (unbinder != null) {
            unbinder.unbind();
            unbinder = null;
        }
    }

    @Override
    public void onDismiss(DialogInterface dialog) {
        super.onDismiss(dialog);

        FragmentActivity activity = getActivity();
        if (activity instanceof DialogDismissListener) {
            // propagate to fragment
            ((DialogDismissListener) activity).onDialogDismissed(this);
        }
    }

    protected Context getThemedContext() {
        Dialog dialog = getDialog();
        return dialog != null ? dialog.getContext() : getContext();
    }

    @Override
    public void dismiss() {
        try {
            super.dismiss();
        } catch (Exception ignored) {
            // i never want that!
        }
    }

    @Override
    public void dismissAllowingStateLoss() {
        try {
            super.dismissAllowingStateLoss();
        } catch (Exception ignored) {
            // i never want that!
        }
    }

    public void dismissNow() {
        try {
            dismiss();
            getFragmentManager().executePendingTransactions();
        } catch (Exception err) {
            // woot!?
        }
    }
}

File: app/src/main/java/com/pr0gramm/app/ui/base/BaseDialogFragment.kt
package com.pr0gramm.app.ui.base

import android.app.Dialog
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import androidx.appcompat.app.AppCompatDialogFragment
import androidx.appcompat.view.ContextThemeWrapper
import androidx.viewbinding.ViewBinding
import com.pr0gramm.app.Logger
import com.pr0gramm.app.time
import com.pr0gramm.app.ui.dialogs.DialogDismissListener
import com.pr0gramm.app.ui.resolveDialogTheme
import com.pr0gramm.app.util.di.LazyInjectorAware
import com.pr0gramm.app.util.di.PropertyInjector

/**
 * A robo fragment that provides lifecycle events as an observable.
 */
abstract class BaseDialogFragment(name: String) : AppCompatDialogFragment(), LazyInjectorAware {
    protected val logger = Logger(name)

    override val injector: PropertyInjector = PropertyInjector()

    override fun onCreate(savedInstanceState: Bundle?) {
        logger.time("Injecting services") { injector.inject(requireContext()) }
        super.onCreate(savedInstanceState)
    }

    final override fun onStart() {
        super.onStart()

        // bind dialog. It is only created in on start.
        dialog?.let { onDialogViewCreated() }
    }

    protected open fun onDialogViewCreated() {}

    override fun onDismiss(dialog: DialogInterface) {
        super.onDismiss(dialog)

        val activity = activity
        if (activity is DialogDismissListener) {
            // propagate to fragment
            activity.onDialogDismissed(this)
        }
    }

    override fun dismiss() {
        try {
            super.dismiss()
        } catch (ignored: Exception) {
            // i never want that!
        }
    }

    override fun dismissAllowingStateLoss() {
        try {
            super.dismissAllowingStateLoss()
        } catch (ignored: Exception) {
            // i never want that!
        }
    }
}


abstract class ViewBindingDialogFragment<T : ViewBinding>(
        name: String, private val inflate: (inflater: LayoutInflater) -> T,
) : BaseDialogFragment(name) {

    protected lateinit var views: T

    final override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        // apply default dialog theme to context before inflating the
        // bindings, so that the dialog theme is applied to the view.
        val themedContext = ContextThemeWrapper(
                requireContext(), resolveDialogTheme(requireContext(), theme),
        )

        views = inflate(LayoutInflater.from(themedContext))
        return onCreateDialog(views.root)
    }

    abstract fun onCreateDialog(contentView: View): Dialog
}